### DISCUSSION TITLE
Pyinstaller Dependency

### DISCUSSION THREAD

**Discussion by roidelfin** [pyinstaller/pyinstaller#8336] [PyInstaller]
I have a problem with my pyinstaller export. I get the following error. 

1187 INFO: PyInstaller: 6.4.0, contrib hooks: 2024.1
1187 INFO: Python: 3.12.2
1219 INFO: Platform: Windows-11-10.0.22631-SP0
1219 INFO: wrote C:\Users\roide\Desktop\Virtual Environment\Final EXE File\daily.spec
1219 INFO: Extending PYTHONPATH with paths
['C:\\Users\\roide\\Desktop\\Virtual Environment\\Final EXE File']
1549 INFO: checking Analysis
1549 INFO: Building Analysis because Analysis-00.toc is non existent
1551 INFO: Initializing module dependency graph...
1551 INFO: Caching module graph hooks...
1551 INFO: Analyzing base_library.zip ...
2878 INFO: Loading module hook 'hook-heapq.py' from 'C:\\Users\\roide\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\PyInstaller\\hooks'...
2957 INFO: Loading module hook 'hook-encodings.py' from 'C:\\Users\\roide\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\PyInstaller\\hooks'...
4469 INFO: Loading module hook 'hook-pickle.py' from 'C:\\Users\\roide\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\PyInstaller\\hooks'...
5495 INFO: Caching module dependency graph...
5574 INFO: Running Analysis Analysis-00.toc
5574 INFO: Looking for Python shared library...
5589 INFO: Using Python shared library: C:\Users\roide\AppData\Local\Programs\Python\Python312\python312.dll
5589 INFO: Analyzing C:\Users\roide\Desktop\Virtual Environment\Final EXE File\daily.py
5589 INFO: Processing module hooks...
5589 INFO: Performing binary vs. data reclassification (2 entries)
5589 INFO: Looking for ctypes DLLs
5605 INFO: Analyzing run-time hooks ...
5605 INFO: Including run-time hook 'C:\\Users\\roide\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\PyInstaller\\hooks\\rthooks\\pyi_rth_inspect.py'
5605 INFO: Looking for dynamic libraries
5731 INFO: Extra DLL search directories (AddDllDirectory): []
5731 INFO: Extra DLL search directories (PATH): []
5954 WARNING: Library not found: could not resolve 'api-ms-win-core-path-l1-1-0.dll', dependency of 'C:\\Users\\roide\\AppData\\Local\\Programs\\Python\\Python312\\python312.dll'.
5954 INFO: Warnings written to C:\Users\roide\Desktop\Virtual Environment\Final EXE File\build\daily\warn-daily.txt
5969 INFO: Graph cross-reference written to C:\Users\roide\Desktop\Virtual Environment\Final EXE File\build\daily\xref-daily.html
5985 INFO: checking PYZ
5985 INFO: Building PYZ because PYZ-00.toc is non existent
5985 INFO: Building PYZ (ZlibArchive) C:\Users\roide\Desktop\Virtual Environment\Final EXE File\build\daily\PYZ-00.pyz
6220 INFO: Building PYZ (ZlibArchive) C:\Users\roide\Desktop\Virtual Environment\Final EXE File\build\daily\PYZ-00.pyz completed successfully.
6258 INFO: checking PKG
6258 INFO: Building PKG because PKG-00.toc is non existent
6258 INFO: Building PKG (CArchive) daily.pkg
12344 INFO: Building PKG (CArchive) daily.pkg completed successfully.
12351 INFO: Bootloader C:\Users\roide\AppData\Local\Programs\Python\Python312\Lib\site-packages\PyInstaller\bootloader\Windows-64bit-intel\run.exe
12351 INFO: checking EXE
12351 INFO: Building EXE because EXE-00.toc is non existent
12351 INFO: Building EXE from EXE-00.toc
12351 INFO: Copying bootloader EXE to C:\Users\roide\Desktop\Virtual Environment\Final EXE File\dist\daily.exe
12363 INFO: Copying icon to EXE
12372 INFO: Copying 0 resources to EXE
12373 INFO: Embedding manifest in EXE
12382 INFO: Appending PKG archive to EXE
12392 INFO: Fixing EXE headers
12580 INFO: Building EXE from EXE-00.toc completed successfully.

**Comment by rokm** [comment]
> I have a problem with my pyinstaller export. I get the following error.

So what is the problem and where is the error? I only see a non-fatal warning there.

**Reply by roidelfin** [reply]
The exe can't execute properly because of this missing dependency. 

5954 WARNING: Library not found: could not resolve 'api-ms-win-core-path-l1-1-0.dll', dependency of 'C:\Users\roide\AppData\Local\Programs\Python\Python312\python312.dll'.

**Reply by rokm** [reply]
What is the actual *run-time* error?

**Reply by roidelfin** [reply]
When running the compiled exe file, it returns: 

Traceback (most recent call last):
  File "daily.py", line 239, in <module>
NameError: name 'null' is not defined
[14276] Failed to execute script 'daily' due to unhandled exception!

**Reply by rokm** [reply]
What does the line 239 in `daily.py` look like, then?

**Reply by roidelfin** [reply]
"execution_count": null

**Reply by roidelfin** [reply]
I think the issue relies on the "api-ms-win-core-path-l1-1-0.dll".

**Reply by rokm** [reply]
> I think the issue relies on the "api-ms-win-core-path-l1-1-0.dll".

It doesn't. That warning should be suppressed, and the only reason you're seeing it is because we suppress it when building on Windows 10 but not on 11, which should be fixed. But like I said, that warning is non-fatal.

> "execution_count": null

>  NameError: name 'null' is not defined

Sounds like that `null` is not defined anywhere. Did you mean to use `None`?

**Reply by roidelfin** [reply]
I have tried replacing null with None and used the same pyinstaller command but it still doesn't work.

**Reply by rokm** [reply]
Is it the exactly same error, or has it changed?

**Reply by roidelfin** [reply]
It no longer returns an error. When I call the .exe file from my command prompt, it will just automatically return to the next line. No error prompts whatsoever.

**Reply by rokm** [reply]
Without seeing your code, I cannot help you then. Maybe you can add print statements to your code to see how far it goes and where the program stops executing.

**Reply by roidelfin** [reply]
This is my python notebook code: 

```
# Import all the libraries
from office365.runtime.auth.authentication_context import AuthenticationContext
from office365.sharepoint.client_context import ClientContext
from office365.sharepoint.files.file import File
import io
import pandas as pd
import psycopg2
from datetime import datetime, timedelta
from openpyxl import Workbook, load_workbook
import numpy as np

# Target SharePoint URL and credentials
url = #'microsoft Sharepoint URL'
username = #'microsoft username'
password = #'microsoft password'

try:
    ctx_auth = AuthenticationContext(url)
    if ctx_auth.acquire_token_for_user(username, password):
        ctx = ClientContext(url, ctx_auth)
        web = ctx.web
        ctx.load(web)
        ctx.execute_query()
        print("Authentication successful")
        
        # To ensure yesterday's data is obtained
        # Returns month name in string
        month = (datetime.today() - timedelta(days = 1)).strftime("%b") + " 2024"
        
        response = File.open_binary(ctx, f"Sharepoint URL.xlsx")

        # Save data to BytesIO stream
        bytes_file_obj = io.BytesIO()
        bytes_file_obj.write(response.content)
        bytes_file_obj.seek(0)  # Set file object to start
        
        # Function to export data from Excel and insert it into PostgreSQL
        def export_to_postgresql(excel_file, db_config):
            
            conn = psycopg2.connect(**db_config)
            cursor = conn.cursor()
            
            excel_data = pd.ExcelFile(excel_file)
            sheet_names = excel_data.sheet_names
            wb = load_workbook(excel_file)
            
            # Select the last sheet (assuming it's the latest created)
            latest_sheet_name = sheet_names[0]

            # Read excel file and each sheet into pandas dataframe
            df = pd.read_excel(bytes_file_obj, sheet_name=latest_sheet_name, skiprows=6, nrows=120, usecols="E:J")
            ws = wb[latest_sheet_name]
            iso = ws['B7'].value if ws['B7'].value else ws['C7'].value if ws['C7'].value else np.nan
            df.loc[len(df.index)] = np.repeat(iso, len(df.columns))
            df.loc[len(df.index)] = np.repeat(pd.to_datetime(iso[-8:],yearfirst=True).strftime('%Y-%m-%d'), len(df.columns))
            df_transposed = df.transpose()

            # Iterate through the rows of the transposed DataFrame and insert into the database
            for index, row in df_transposed.iterrows():
                values = tuple(row)  # Convert the row to a tuple
                insert_query = f"""
                    INSERT INTO {table_name} (
                        my column names
                    ) VALUES %s
                """
                cursor.execute(insert_query, (values,))
            
            print(f"Data has been successfully inserted into {table_name} table {latest_sheet_name} sheet.")
            
            # Commit the changes and close the cursor and connection
            conn.commit()
            cursor.close()
            conn.close()
            
    else:
        print("Failed to acquire token for user.")
except Exception as e:
    print(f"An error occurred: {str(e)}")

# Configuration for PostgreSQL database
db_config = {
    'host': 'localhost',
    'port': 5432,
    'database': 'postgres',
    'user': 'postgres',
    'password': 'password'
}

# Call the function with your Excel file and database config
excel_file = bytes_file_obj
table_name = 'water_treatment'

# Export function
export_to_postgresql(excel_file, db_config)
```

I then saved my Jupyter notebook to .py so that I can convert it to exe using pyinstaller and this is my .py conversion code: 
```
{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7ee011c1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Authentication successful\n",
      "Data has been successfully inserted into water_treatment table February (29) sheet.\n"
     ]
    }
   ],
   "source": [
    "# Import all the libraries\n",
    "from office365.runtime.auth.authentication_context import AuthenticationContext\n",
    "from office365.sharepoint.client_context import ClientContext\n",
    "from office365.sharepoint.files.file import File\n",
    "import io\n",
    "import pandas as pd\n",
    "#import glob\n",
    "import psycopg2\n",
    "from datetime import datetime, timedelta\n",
    "from openpyxl import Workbook, load_workbook\n",
    "import numpy as np\n",
    "\n",
    "# Target SharePoint URL and credentials\n",
    "url = '",
    "username = '",
    "password = '",
    "\n",
    "try:\n",
    "    ctx_auth = AuthenticationContext(url)\n",
    "    if ctx_auth.acquire_token_for_user(username, password):\n",
    "        ctx = ClientContext(url, ctx_auth)\n",
    "        web = ctx.web\n",
    "        ctx.load(web)\n",
    "        ctx.execute_query()\n",
    "        print(\"Authentication successful\")\n",
    "        \n",
    "        # To ensure yesterday's data is obtained\n",
    "        # Returns month name in string\n",
    "        month = (datetime.today() - timedelta(days = 1)).strftime(\"%b\") + \" 2024\"\n",
    "        \n",
    "        response = File.open_binary(ctx, f\"",
    "\n",
    "        # Save data to BytesIO stream\n",
    "        bytes_file_obj = io.BytesIO()\n",
    "        bytes_file_obj.write(response.content)\n",
    "        bytes_file_obj.seek(0)  # Set file object to start\n",
    "        \n",
    "        # Function to export data from Excel and insert it into PostgreSQL\n",
    "        def export_to_postgresql(excel_file, db_config):\n",
    "            \n",
    "            conn = psycopg2.connect(**db_config)\n",
    "            cursor = conn.cursor()\n",
    "            \n",
    "            excel_data = pd.ExcelFile(excel_file)\n",
    "            sheet_names = excel_data.sheet_names\n",
    "            wb = load_workbook(excel_file)\n",
    "            \n",
    "            # Select the last sheet (assuming it's the latest created)\n",
    "            latest_sheet_name = sheet_names[0]\n",
    "\n",
    "            # Read excel file and each sheet into pandas dataframe\n",
    "            df = pd.read_excel(bytes_file_obj, sheet_name=latest_sheet_name, skiprows=6, nrows=120, usecols=\"E:J\")\n",
    "            ws = wb[latest_sheet_name]\n",
    "            iso = ws['B7'].value if ws['B7'].value else ws['C7'].value if ws['C7'].value else np.nan\n",
    "            df.loc[len(df.index)] = np.repeat(iso, len(df.columns))\n",
    "            df.loc[len(df.index)] = np.repeat(pd.to_datetime(iso[-8:],yearfirst=True).strftime('%Y-%m-%d'), len(df.columns))\n",
    "            df_transposed = df.transpose()\n",
    "\n",
    "            # Iterate through the rows of the transposed DataFrame and insert into the database\n",
    "            for index, row in df_transposed.iterrows():\n",
    "                values = tuple(row)  # Convert the row to a tuple\n",
    "                insert_query = f\"\"\"\n",
    "                    INSERT INTO {table_name} (\n",
    "                       column names
    "                    ) VALUES %s\n",
    "                \"\"\"\n",
    "                cursor.execute(insert_query, (values,))\n",
    "            \n",
    "            print(f\"Data has been successfully inserted into {table_name} table {latest_sheet_name} sheet.\")\n",
    "            \n",
    "            # Commit the changes and close the cursor and connection\n",
    "            conn.commit()\n",
    "            cursor.close()\n",
    "            conn.close()\n",
    "            \n",
    "    else:\n",
    "        print(\"Failed to acquire token for user.\")\n",
    "except Exception as e:\n",
    "    print(f\"An error occurred: {str(e)}\")\n",
    "\n",
    "# Configuration for PostgreSQL database\n",
    "db_config = {\n",
    "    'host': 'localhost',\n",
    "    'port': 5432,\n",
    "    'database': 'postgres',\n",
    "    'user': 'postgres',\n",
    "    'password': 'password'\n",
    "}\n",
    "\n",
    "# Call the function with your Excel file and database config\n",
    "excel_file = bytes_file_obj\n",
    "table_name = 'water_treatment'\n",
    "\n",
    "# Export function\n",
    "export_to_postgresql(excel_file, db_config)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "889234d6-8e2b-423d-8870-1c0bb2b93d79",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
```

**Reply by roidelfin** [reply]
I am using Windows 11.

**Reply by rokm** [reply]
So if you build this program and run it from command prompt, do you see any output at all?

**Reply by roidelfin** [reply]
I have already solved the issue. Thanks!
